File: MTGSearch/src/test/java/com/dbottillo/mtgsearchfree/presenter/CardPresenterImplTest.java
package com.dbottillo.mtgsearchfree.presenter;

import com.dbottillo.mtgsearchfree.interactors.CardsInteractor;
import com.dbottillo.mtgsearchfree.model.MTGCard;
import com.dbottillo.mtgsearchfree.util.Logger;
import com.dbottillo.mtgsearchfree.view.CardView;

import org.junit.Before;
import org.junit.Rule;
import org.junit.Test;
import org.mockito.ArgumentCaptor;
import org.mockito.Mock;
import org.mockito.junit.MockitoJUnit;
import org.mockito.junit.MockitoRule;

import io.reactivex.Observable;

import static org.hamcrest.CoreMatchers.is;
import static org.junit.Assert.assertThat;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

public class CardPresenterImplTest {

    @Rule
    public MockitoRule mockitoRule = MockitoJUnit.rule();

    private CardPresenter underTest;

    @Mock
    CardsInteractor cardsInteractor;

    @Mock
    Logger logger;

    @Mock
    CardView cardView;

    @Mock
    MTGCard card;

    @Mock
    MTGCard otherCard;

    @Before
    public void setUp() throws Exception {
        when(cardsInteractor.loadOtherSideCard(card)).thenReturn(Observable.just(otherCard));
        underTest = new CardPresenterImpl(cardsInteractor, logger, new TestRunnerFactory());
        underTest.init(cardView);
    }

    @Test
    public void loadCardWithMultiverseId() {
        underTest.loadOtherSideCard(card);

        verify(cardsInteractor).loadOtherSideCard(card);
        ArgumentCaptor<MTGCard> argument = ArgumentCaptor.forClass(MTGCard.class);
        verify(cardView).otherSideCardLoaded(argument.capture());
        assertThat(argument.getValue(), is(otherCard));
    }

}

File: MTGSearch/src/test/java/com/dbottillo/mtgsearchfree/view/views/CardPresenterImplTest.kt
package com.dbottillo.mtgsearchfree.view.views

import com.dbottillo.mtgsearchfree.interactors.CardsInteractor
import com.dbottillo.mtgsearchfree.model.MTGCard
import com.dbottillo.mtgsearchfree.util.Logger
import com.dbottillo.mtgsearchfree.view.CardView
import com.dbottillo.mtgsearchfree.view.views.CardPresenter
import com.dbottillo.mtgsearchfree.view.views.CardPresenterImpl

import org.junit.Before
import org.junit.Rule
import org.junit.Test
import org.mockito.ArgumentCaptor
import org.mockito.Mock
import org.mockito.junit.MockitoJUnit
import org.mockito.junit.MockitoRule

import io.reactivex.Observable

import org.hamcrest.CoreMatchers.`is`
import org.junit.Assert.assertThat
import org.mockito.Mockito.*

class CardPresenterImplTest {

    @Rule @JvmField
    var mockitoRule = MockitoJUnit.rule()

    lateinit var underTest: CardPresenter

    @Mock
    lateinit var cardsInteractor: CardsInteractor

    @Mock
    lateinit var logger: Logger

    @Mock
    lateinit var view: CardView

    @Mock
    lateinit var card: MTGCard

    @Mock
    lateinit var otherCard: MTGCard

    @Before
    fun setUp() {
        underTest = CardPresenterImpl(cardsInteractor, logger)
        underTest.init(view)
    }

    @Test
    fun `load other side card should call interactor and update view`() {
        `when`(cardsInteractor.loadOtherSideCard(card)).thenReturn(Observable.just(otherCard))
        
        underTest.loadOtherSideCard(card)

        verify(cardsInteractor).loadOtherSideCard(card)
        verify(view).otherSideCardLoaded(otherCard)
        verifyNoMoreInteractions(cardsInteractor, view)
    }

}